pair/src/lib.rs

```diff
@@ -130,97 +253,158 @@ impl Pair {
         self.balance0 += amount0;
         self.balance1 += amount1;
         // call mint function
-        let liquidity = self.mint(to).await;
-        msg::reply(
-            PairEvent::AddedLiquidity {
-                amount0,
-                amount1,
-                liquidity,
-                to,
+        let liquidity = self.mint(*to).await;
+
+        // now we can transfer tokens
+        self.perform_two_transfers(
+            transaction_id,
+            transaction_hash,
+            TransferDescription {
+                token_address: self.token0,
+                from: msg::source(),
+                to: pair_address,
+                token_amount: amount0,
+            },
+            TransferDescription {
+                token_address: self.token1,
+                from: msg::source(),
+                to: pair_address,
+                token_amount: amount1,
             },
-            0,
         )
-        .expect("Error during a replying with PairEvent::AddedLiquidity");
+        .await;
     }
 
-    /// Removes liquidity from the pool.
-    /// Internally calls self.burn function while transferring `liquidity` amount of internal tokens
-    /// `to` - MUST be a non-zero address
-    /// Arguments:
-    /// * `liquidity` - is the desired liquidity the user wants to remove (e.g. burn)
-    /// * `amount0_min` - is the minimum amount of token0 the user wants to receive
-    /// * `amount1_min` - is the minimum amount of token1 the user wants to receive
-    /// * `to` - is the liquidity provider
-    pub async fn remove_liquidity(
+    async fn remove_liquidity(
         &mut self,
+        transaction_id: u64,
         liquidity: u128,
         amount0_min: u128,
         amount1_min: u128,
-        to: ActorId,
+        to: &ActorId,
     ) {
+        let transaction_hash = get_hash(&msg::source(), transaction_id);
         FTCore::transfer(self, &msg::source(), &exec::program_id(), liquidity);
-        // Burn and get the optimal amount of burned tokens.
-        let (amount0, amount1) = self.burn(to).await;
+        // no need for self.burn though
+        let fee_on = self.mint_fee(self.reserve0, self.reserve1).await;
+        let liquidity: u128 = *self
+            .get()
+            .balances
+            .get(&exec::program_id())
+            .expect("The pair has no liquidity at all");
+        let amount0 = liquidity
+            .wrapping_mul(self.balance0)
+            .wrapping_div(self.get().total_supply);
+        let amount1 = liquidity
+            .wrapping_mul(self.balance1)
+            .wrapping_div(self.get().total_supply);
+
+        if amount0 == 0 || amount1 == 0 {
+            panic!("PAIR: Insufficient liquidity burnt.");
+        }
+        self.update_balance(*to, liquidity, false);
+        self.balance0 -= amount0;
+        self.balance1 -= amount1;
+        self.update(self.balance0, self.balance1, self.reserve0, self.reserve1);
```

Comment:
`self` method with all `self.*` params looks weird.
```suggestion
        self.update();
```